### ISSUE TITLE
.dockerignore is ignored and details like .git/config with credentials are added to images

### ISSUE THREAD

**Issue by raisercostin** [caprover/caprover#2157]
# What is the problem?
When deploying using Method 3: Deploy from Github/Bitbucket/Gitlab `.dockerignore` file is ignored.
The expected behaviour is that any docker build process using `Dockerfile` will also use `.dockerignore` if is present.
I've created a minimalist project at https://github.com/raisercostin/caprover-test-dockerignore.git
See also https://github.com/caprover/caprover/issues/787#issuecomment-1436742525

# Steps to reproduce the problem:
1. create an app and add in method3: https://github.com/raisercostin/caprover-test-dockerignore.git main any user and password
2. force build
3. in the website /.git/config should not exist

# Output of the following command on your server:
```
Linux AMANTAWIN3 5.15.153.1-microsoft-standard-WSL2 #1 SMP Fri Mar 29 23:14:13 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.4 LTS
Release:        22.04
Codename:       jammy
               total        used        free      shared  buff/cache   available
Mem:            15Gi       835Mi        13Gi       3.0Mi       1.1Gi        14Gi
Swap:          4.0Gi          0B       4.0Gi
```

**Comment by maietta** [comment]
I don't know your setup, so I can only speculate.

There are two types of models for deploying to a system like CapRover.

A pull type method, or a push type method. It sounds like you might be pushing the files to let CapRover build. If this is the case, you need to make sure to include the .dockerignore file as part of the list of files and directories to copy over for CapRover do finish deploying from.

If it's a "pull" situation from CapRover's perspective, then it sounds like the .dockerignore wasn't even there in the files to begin with.

**Comment by raisercostin** [comment]
More details: the deployment is like this using the shared minimalistic project https://github.com/raisercostin/caprover-test-dockerignore.git

![image](https://github.com/user-attachments/assets/244b710d-2cdf-415a-89ed-080ad7c22633)


So the `index.html` file and all others (including .git) get to be added to image despite `.dockerignore` saying otherwise.
```
.git
.gitignore
Dockerfile
```

**Comment by githubsaturn** [comment]
Does `caprover deploy` work fine?

**Comment by githubsaturn** [comment]
Was able to simply reproduce even without CapRover.

See https://github.com/caprover/caprover/issues/2157

- Clone this repo
- Run `rm -rf .git` to ensure this repo's git is deleted
- Run `sudo ./build_image.sh`

You will see that the `.git` directory (inside the tarfile) was copied over despite of .dockerignore.

**Comment by githubsaturn** [comment]
Looks like Docker API completely ignores the `.dockerignore` file even though Docker CLI respects this file.

**Comment by raisercostin** [comment]
> Does `caprover deploy` work fine?

Sorry I'm not familiar with this type of deployment. But I understand that you reproduced the problem. Thank you.

**Comment by githubsaturn** [comment]
Yeah it looks like the problem is not with CapRover. Rather it's a problem with Docker API. I'll open a ticket with Docker to have them address this issue.

**Comment by Riches** [comment]
I would just like to highlight this issue again. I've just stumbled across this issue after finding unexpected files in my images.

From my perspective this is an unexpected behaviour verging on a security issue. For example, files containing secrets may be in your .dockerignore but these are actually ending up in the image and ultimately the live containers. And of course these containers _could_ be pushed to public registries (outside the scope of CapRover I know) which would be even worse.

Docker developers will likely be familiar in the behaviour of .dockerignore with the Docker CLI, but not this more complex behaviour via the Docker API. At a minimum, I would suggest documenting this behaviour very clearly. Especially as Docker says there's no bug to fix and this is expected behaviour, so I'm not sure if a fix would even be possible.

Brilliant software otherwise @githubsaturn !

**Comment by githubsaturn** [comment]
Agreed. Unfortunately Docker's maintainers closed this as "working as expected". There is little we can do to fix this other than documenting this behavior. I'd like to know where you expect this information to exist? Perhaps somewhere in [this page](https://caprover.com/docs/captain-definition-file.html)?
